crates/subspace-service/src/domains.rs

```diff
@@ -0,0 +1,229 @@
+
+pub struct LastDomainBlockInfoReceiver<Block, Client, NR>
+where
+    Block: BlockT,
+    NR: NetworkRequest,
+    Client: HeaderBackend<Block>,
+{
+    domain_id: DomainId,
+    fork_id: Option<String>,
+    client: Arc<Client>,
+    network_service: NR,
+    sync_service: Arc<SyncingService<Block>>,
+}
+
+impl<Block, Client, NR> LastDomainBlockInfoReceiver<Block, Client, NR>
+where
+    Block: BlockT,
+    NR: NetworkRequest,
+    Client: HeaderBackend<Block>,
+{
+    pub fn new(
+        domain_id: DomainId,
+        fork_id: Option<String>,
+        client: Arc<Client>,
+        network_service: NR,
+        sync_service: Arc<SyncingService<Block>>,
+    ) -> Self {
+        Self {
+            domain_id,
+            fork_id,
+            client,
+            network_service,
+            sync_service,
+        }
+    }
+    pub async fn get_last_confirmed_domain_block_receipt<CBlock: BlockT>(
+        &self,
+        block_hash: Option<CBlock::Hash>,
+    ) -> Option<ExecutionReceiptFor<Block::Header, CBlock, Balance>> {
+        let info = self.client.info();
+        let protocol_name = generate_protocol_name(info.genesis_hash, self.fork_id.as_deref());
+
+        debug!(domain_id=%self.domain_id, %protocol_name, "Started obtaining domain info...");
+
+        loop {
+            let peers_info = match self.sync_service.peers_info().await {
+                Ok(peers_info) => peers_info,
+                Err(error) => {
+                    error!("Peers info request returned an error: {error}",);
+                    sleep(REQUEST_PAUSE).await;
+
+                    continue;
+                }
+            };
+
+            //  Enumerate peers until we find a suitable source for domain info
+            'peers: for (peer_id, peer_info) in peers_info.iter() {
+                debug!(
+                    "Domain data request. peer = {peer_id}, info = {:?}",
+                    peer_info
+                );
+
+                if !peer_info.is_synced {
+                    trace!("Domain data request skipped (not synced). peer = {peer_id}");
+
+                    continue;
+                }
+
+                let request = LastConfirmedBlockRequest::<CBlock> {
+                    domain_id: self.domain_id,
+                    block_hash,
+                };
+
+                let response = send_request::<NR, CBlock, Block::Header>(
+                    protocol_name.clone(),
+                    *peer_id,
+                    request,
+                    &self.network_service,
+                )
+                .await;
+
+                match response {
+                    Ok(response) => {
+                        trace!("Response from a peer {peer_id},",);
+
+                        return Some(response.last_confirmed_block_receipt);
+                    }
+                    Err(error) => {
+                        debug!("Domain info request failed. peer = {peer_id}: {error}");
+
+                        continue 'peers;
```

Comment:
Nitpick: Is the `'peers` label needed?

I think it's redundant here (and above), because there's only one loop in this method:
https://doc.rust-lang.org/reference/expressions/loop-expr.html#continue-expressions

Reply:
There is another external unconditional loop.

Reply:
Thank you for explaining, I see that now.

But the label is still redundant, because `continue` already applies to the innermost loop in Rust.
So should there be a label going to the outermost loop at line 141 or 166?
Or should there be a redundant label on both lines pointing to the innermost loop for clarity?

Reply:
>Or should there be a redundant label on both lines pointing to the innermost loop for clarity?

IMHO, when reading nested loop code it is helpful to have a label to clarify which loop the `continue/break` applies to, even though the label is redundant. Non-blocker.

Reply:
Great, then let's do that consistently?
https://github.com/autonomys/subspace/pull/2999/files#r1749398297